skimage/morphology/_flood_fill.py

```diff
@@ -0,0 +1,290 @@
+        If None (default), adjacent values must be strictly equal to the
+        initial value of `image` at `seed_point`.  This is fastest.  If a value
+        is given, a comparison will be done at every point and this tolerance
+        on each side of the initial value will also be filled (inclusive).
+    inplace : bool, optional
+        If True, flood filling is applied to `image` inplace.  If False, the
+        flood filled result is returned without modifying the input `image`
+        (default).
+
+    Returns
+    -------
+    filled : ndarray
+        An array with the same shape as `image` is returned with values equal
+        to (or within tolerance of) the seed point set to `new_value`.
+
+    Notes
+    -----
+    The conceptual analogy of this operation is the 'paint bucket' tool in many
+    raster graphics programs.
+
+    Examples
+    --------
+    >>> from skimage.morphology import flood_fill
+    >>> image = np.zeros((4, 7), dtype=int)
+    >>> image[1:3, 1:3] = 1
+    >>> image[3, 0] = 1
+    >>> image[1:3, 4:6] = 2
+    >>> image[3, 6] = 3
+    >>> image
+    array([[0, 0, 0, 0, 0, 0, 0],
+           [0, 1, 1, 0, 2, 2, 0],
+           [0, 1, 1, 0, 2, 2, 0],
+           [1, 0, 0, 0, 0, 0, 3]])
+
+    Fill connected ones with 5, with full connectivity (diagonals included):
+
+    >>> flood_fill(image, (1, 1), 5)
+    array([[0, 0, 0, 0, 0, 0, 0],
+           [0, 5, 5, 0, 2, 2, 0],
+           [0, 5, 5, 0, 2, 2, 0],
+           [5, 0, 0, 0, 0, 0, 3]])
+
+    Fill connected ones with 5, with only cardinal direction connectivity:
+
+    >>> flood_fill(image, (1, 1), 5, connectivity=1)
+    array([[0, 0, 0, 0, 0, 0, 0],
+           [0, 5, 5, 0, 2, 2, 0],
+           [0, 5, 5, 0, 2, 2, 0],
+           [1, 0, 0, 0, 0, 0, 3]])
+
+    Fill with a tolerance:
+
+    >>> flood_fill(image, (0, 0), 5, tolerance=1)
+    array([[5, 5, 5, 5, 5, 5, 5],
+           [5, 5, 5, 5, 2, 2, 5],
+           [5, 5, 5, 5, 2, 2, 5],
+           [5, 5, 5, 5, 5, 5, 3]])
+    """
+    # Correct start point in ravelled image - only copy if non-contiguous
+    image = np.asarray(image)
+    copied = False
+
+    if not image.flags.contiguous:
+        if inplace:
+            warnings.warn('Non-contiguous array passed as `image`; this will '
+                          'be converted to a contiguous array as a copy.')
+            # Flood fill requires some sort of contiguity - this makes a copy
+            image = np.ascontiguousarray(image)
+            copied = True
+
+    if not inplace:
+        if not copied:
+            output = image.copy()
+        else:
+            output = image
+        output[flood(image, seed_point, selem=selem, connectivity=connectivity,
+                     tolerance=tolerance)] = new_value
+        return output
+    else:  # inplace
+        image[flood(image, seed_point, selem=selem, connectivity=connectivity,
+                    tolerance=tolerance)] = new_value
+        return image
```

Comment:
I find the nested workflow above confusing to read at a first glance. Maybe there is a simpler solution? Wouldn't

```python
if inplace and not image.flags.contiguous:
    warnings.warn('Non-contiguous array passed as `image`; this will '
                  'be converted to a contiguous array as a copy.')
    # Flood fill requires some sort of contiguity - this makes a copy
    image = np.ascontiguousarray(image)
elif not inplace:
    image = image.copy()

mask = flood(image, seed_point, selem=selem, connectivity=connectivity,
             tolerance=tolerance)
image[mask] = new_value
return image
```

do the same?

Reply:
Wait, I think we don't even need to check for contiguity here because `flood` will do that for us. `image[mask] = new_value` should work for non-contiguous arrays just fine. The only case where this would bite us is if we pass in a non-contiguous image with `inplace=False` in which case the image would be copied twice (once in `flood_fill` and once in `flood`). So maybe having a check for contiguity in both functions is for the best.

Reply:
I think @lagru has uncovered a valid concern here. As far as I can tell, `inplace=True` won't work for non-contiguous arrays, because `image` will have been copied. However this is not necessary because `image[mask] = new_value` will work fine for noncontiguous, so even though the Cython code requires contiguity, we can still do the fill in-place.

Am I wrong here? And, if not, @JDWarner can you change this behaviour and add a test for it?

Reply:
@jni Yeah, that's what I meant. (I modified my original snippet.)

Reply:
Done.